test/framework/src/main/java/org/elasticsearch/test/transport/MockTransport.java

```diff
@@ -0,0 +1,268 @@
+        Tuple<DiscoveryNode, String> request = requests.get(requestId);
+        assert request != null;
+        this.handleError(requestId, new SendRequestTransportException(request.v1(), request.v2(), t));
+    }
+
+    /**
+     * simulate a remote error for the given requestId, will be wrapped
+     * by a {@link RemoteTransportException}
+     *
+     * @param requestId the id corresponding to the captured send
+     *                  request
+     * @param t         the failure to wrap
+     */
+    public void handleRemoteError(final long requestId, final Throwable t) {
+        final RemoteTransportException remoteException;
+        if (rarely(Randomness.get())) {
+            remoteException = new RemoteTransportException("remote failure, coming from local node", t);
+        } else {
+            try (BytesStreamOutput output = new BytesStreamOutput()) {
+                output.writeException(t);
+                remoteException = new RemoteTransportException("remote failure", output.bytes().streamInput().readException());
+            } catch (IOException ioException) {
+                throw new ElasticsearchException("failed to serialize/deserialize supplied exception " + t, ioException);
+            }
+        }
+        this.handleError(requestId, remoteException);
+    }
+
+    /**
+     * simulate an error for the given requestId, unlike
+     * {@link #handleLocalError(long, Throwable)} and
+     * {@link #handleRemoteError(long, Throwable)}, the provided
+     * exception will not be wrapped but will be delivered to the
+     * transport layer as is
+     *
+     * @param requestId the id corresponding to the captured send
+     *                  request
+     * @param e         the failure
+     */
+    public void handleError(final long requestId, final TransportException e) {
+        responseHandlers.onResponseReceived(requestId, listener).handleException(e);
+    }
+
+    @Override
+    public Connection openConnection(DiscoveryNode node, ConnectionProfile profile) {
+        return new Connection() {
+            @Override
+            public DiscoveryNode getNode() {
+                return node;
+            }
+
+            @Override
+            public void sendRequest(long requestId, String action, TransportRequest request, TransportRequestOptions options)
+                throws TransportException {
+                onSendRequest(requestId, action, request, node);
+            }
+
+            @Override
+            public void addCloseListener(ActionListener<Void> listener) {
+            }
+
+            @Override
+            public boolean isClosed() {
+                return false;
+            }
+
+            @Override
+            public void close() {
+            }
+        };
+    }
+
+    protected void onSendRequest(long requestId, String action, TransportRequest request, DiscoveryNode node) {
+        requests.put(requestId, Tuple.tuple(node, action));
```

Comment:
All the implementations call `super(...)` up front, and I think it'd be bad not to, so I think it'd be better to move this into `sendRequest()` above.